### PR TITLE
Add and apply "advisory rules" to overwrite targeted version constraints

### PR DESCRIPTION
**Pull request by Fahl-Design**
As seen in Roave/SecurityAdvisories#90 and https://github.com/Roave/SecurityAdvisories/commit/fa60732bcafd446e9ff858eefa06cdcbae8d9823#commitcomment-65800620 there is a need to provide a way to fix some version constraints.
And In addition to the patch request: https://github.com/Roave/SecurityAdvisories/issues/90#issuecomment-1028839258 I created a small PoC (with unit tests)

any input is welcome

### CODE REVIEW

**Comment by alexislefebvre** [comment]
How will people have to use it? I think a documentation and/or example would be useful:
- how to add a rule to override a "bad" rule reported by GitHub
- where must the file go
- does the file has to been removed once GitHub fixed the rules

**Comment by Fahl-Design** [comment]
@alexislefebvre 
> How will people have to use it? I think a documentation and/or example would be useful:

well this is a cli tool to update/sync https://github.com/Roave/SecurityAdvisories run by a workflow action, 

> * how to add a rule to override a "bad" rule reported by GitHub

take a look at https://github.com/Roave/SecurityAdvisoriesBuilder/blob/a6c7cb07470c9dc8fa8876321e8c5c5852d9ac54/src/Roave/SecurityAdvisories/Rule/RuleProviderFactory.php#L26 rules are added there and the provider is used in the cli script

> * where must the file go
> * does the file has to been removed once GitHub fixed the rules

no need to add/remove files. A rule will target a specific constraint in combination to the package name to overwrite it's inital constraints

**Comment by Ocramius** [comment]
I'll probably ask for changes on the PR to keep it as-simple-as-possible, so documentation isn't strictly needed :)

**Comment by alexislefebvre** [comment]
My bad, I thought that `laminas/laminas-form` was the data of a test, but it's a working example. :+1:

File: build-conflicts.php [hunk 1]
```diff
@@ -191,14 +193,22 @@ static function (array $components): array {
...
         (new GetAdvisoriesFromFriendsOfPhp($buildDir . '/security-advisories')),
         (new GetAdvisoriesFromGithubApi(
             new Client(),
             $token,
         )),
     ));
 
+    $overwriteRuleProvider = (new RuleProviderFactory())();
+
+    $getAdvisories = new GetAdvisoriesAdvisoryRuleDecorator(
```
Comment (Ocramius) [inline build-conflicts.php:196]: Let's squash this into a single call, to avoid declaring `$getAdvisoriesSources` (one variable less :) )

File: src/Roave/SecurityAdvisories/Rule/Rule.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * and is licensed under the MIT license.
+ */
+
+declare(strict_types=1);
+
+namespace Roave\SecurityAdvisories\Rule;
+
+use Roave\SecurityAdvisories\Advisory;
+
+interface Rule
```
Comment (Ocramius) [inline src/Roave/SecurityAdvisories/Rule/Rule.php]: IMO interface not needed: let's just make it something as simple as `callable(Advisory): Advisory`, which will (on its own) decide when to replace an advisory entirely.

File: src/Roave/SecurityAdvisories/Rule/AdvisoryRuleProvider.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * and is licensed under the MIT license.
+ */
+
+declare(strict_types=1);
+
+namespace Roave\SecurityAdvisories\Rule;
+
+use Roave\SecurityAdvisories\AdvisoryRule;
+
+final class AdvisoryRuleProvider implements RuleProvider
```
Comment (Ocramius) [inline src/Roave/SecurityAdvisories/Rule/AdvisoryRuleProvider.php]: This can be removed, using a `list<callable(Advisory): Advisory>` instead

File: build-conflicts.php [hunk 4]
```diff
@@ -191,14 +193,22 @@ static function (array $components): array {
...
+    // sources:
+    $getAdvisoriesSources = (new GetAdvisoriesFromMultipleSources(
         (new GetAdvisoriesFromFriendsOfPhp($buildDir . '/security-advisories')),
         (new GetAdvisoriesFromGithubApi(
             new Client(),
             $token,
         )),
     ));
 
+    $overwriteRuleProvider = (new RuleProviderFactory())();
```
Comment (Ocramius) [inline build-conflicts.php]: IMO a file with `list<callable(Advisory): Advisory>` would be simpler here - those callables can be turned into testable classes too

File: build-conflicts.php [hunk 5]
```diff
@@ -191,14 +193,22 @@ static function (array $components): array {
     $cloneAdvisories();
     $cloneRoaveAdvisories();
 
-    $getAdvisories = (new GetAdvisoriesFromMultipleSources(
+    // sources:
```
Comment (Ocramius) [inline build-conflicts.php]: Comment can be dropped

File: src/Roave/SecurityAdvisories/AdvisoryRule.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+namespace Roave\SecurityAdvisories;
+
+use Roave\SecurityAdvisories\Rule\Rule;
+
+final class AdvisoryRule implements Rule
+{
+    /**
+     * @psalm-param array<array-key, array{versions: string|array<array-key, string>}> $branchConfig
+     */
+    public function __construct(
```
Comment (Ocramius) [inline src/Roave/SecurityAdvisories/AdvisoryRule.php]: This syntax is really hard to follow (from a ctor PoV).

IMO, hardcoding these callables for when we have exceptions that require urgent intervention would be better, and easier to maintain

File: src/Roave/SecurityAdvisories/AdvisorySources/GetAdvisoriesAdvisoryRuleDecorator.php [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+use Generator;
+use Roave\SecurityAdvisories\Advisory;
+use Roave\SecurityAdvisories\Rule\Rule;
+use Roave\SecurityAdvisories\Rule\RuleProvider;
+
+final class GetAdvisoriesAdvisoryRuleDecorator implements GetAdvisories
+{
+    public function __construct(
+        private GetAdvisories $source,
+        private RuleProvider $overwriteRuleProvider,
```
Comment (Ocramius) [inline src/Roave/SecurityAdvisories/AdvisorySources/GetAdvisoriesAdvisoryRuleDecorator.php]: `list<callable(Advisory): Advisory>` here too, to cut on the code size :D

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: build-conflicts.php (hunk 5 above)
Comment (Fahl-Design) [reply build-conflicts.php]: done :)

**Comment by Fahl-Design** [comment]
Thx for the review. your are totally right it's much easier to overlook and add/remove rules in a hurry, in this case should change/remove some the unit tests ;) 
I didn't use psalm until yesterday (phpstan was my way of choice) and need a bit more confident in just adding a callable annotation ;) it really looks like a good way to go and dig into (and get my boss to like it too^^)

**Comment by Ocramius** [comment]
Works with phpstan too 👍

File: src/Roave/SecurityAdvisories/Rule/RuleProviderFactory.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                if ($advisory->package->packageName !== $packageName) {
+                    return $advisory;
+                }
+
+                if ($advisory->getConstraint() !== $targetConstraint) {
+                    return $advisory;
+                }
+
+                /**
+                 * @psalm-var array{
```
Comment (Ocramius) [inline src/Roave/SecurityAdvisories/Rule/RuleProviderFactory.php]: Declare it in one shot, and the docblock isn't necessary (let type checks do their job - every time you use `@var` you add fragility)

File: test/RoaveTest/SecurityAdvisories/AdvisorySources/GetAdvisoriesAdvisoryRuleDecoratorTest.php [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+use function method_exists;
+
+/**
+ * Tests for {@see \Roave\SecurityAdvisories\AdvisorySources\GetAdvisoriesAdvisoryRuleDecorator}
+ *
+ * @covers \Roave\SecurityAdvisories\AdvisorySources\GetAdvisoriesAdvisoryRuleDecorator
+ */
+class GetAdvisoriesAdvisoryRuleDecoratorTest extends TestCase
+{
+    public function testThatAdvisoriesAreDecoratedAfterBuiltFromYamlFilesAndConstraintIsChanged(): void
```
Comment (Ocramius) [inline test/RoaveTest/SecurityAdvisories/AdvisorySources/GetAdvisoriesAdvisoryRuleDecoratorTest.php:41]: This test can be simplified massively now: just need to verify that all the given replacement rules are called in sequence, I suppose? 🤔

**Review by Ocramius — CHANGES_REQUESTED** [review]
Massive improvement!

Question: how do we have to adjust each test when the upstream advisories get fixed? Perhaps the callables could be classes that we name based on the advisory/cve name? That means removing the factory, or using the factory to just list them all?

File: src/Roave/SecurityAdvisories/Rule/RuleProviderFactory.php (hunk 8 above)
Comment (Fahl-Design) [reply src/Roave/SecurityAdvisories/Rule/RuleProviderFactory.php]: ah got it

File: test/RoaveTest/SecurityAdvisories/AdvisorySources/GetAdvisoriesAdvisoryRuleDecoratorTest.php (hunk 9 above)
Comment (Fahl-Design) [reply test/RoaveTest/SecurityAdvisories/AdvisorySources/GetAdvisoriesAdvisoryRuleDecoratorTest.php:41]: done, added a rule for each advisory and a dummy rule to verify all rules are called and verify the expected result

**Review by Ocramius — APPROVED** [review]
Excellent, thanks @Fahl-Design!

**Comment by Ocramius** [comment]
All green, 🚢

**Comment by Ocramius** [comment]
@Fahl-Design fyi, didn't seem to work: https://github.com/Roave/SecurityAdvisories/blob/8ba45c14eb7b79db601052ba90c2811d9b028d7a/composer.json#L161 didn't get any changes :| 

Did you try running `build-advisories.php` locally?

**Comment by Fahl-Design** [comment]
hmm I found it
psalm did not catch that one and it can't because it only important in the full context that I missed to ;)

the config is valid in itself:
https://github.com/Roave/SecurityAdvisoriesBuilder/blob/latest/src/Roave/SecurityAdvisories/Rule/RuleProviderFactory.php#L45:L57

but it is always used like this:
https://github.com/Roave/SecurityAdvisoriesBuilder/blob/latest/src/Roave/SecurityAdvisories/AdvisorySources/GetAdvisoriesFromGithubApi.php#L92:L96

and therefor the rule only has to compare that one version,
I will push a fix soon
